arm64-assembly-optimization.md

```diff
@@ -0,0 +1,120 @@
+# Assembly Optimization Guide for Graviton Arm64 Processors
+
+## Introduction
+
+This document is a reference for software developers who want to optimize code
+for Graviton Arm64 processors at the assembly level. Some of the documented
+transforms could be applied to optimize code written in higher level languages
+like C or C++, and in that respect this document is not restrictive to assembly
+level programming.
+
+The code patterns in this document could also be useful to spot inefficient code
+generated for the hot code. Profilers such as Linux perf allow the inspection of
+the assembly level code. During the review of the hot code, software developers
+could reference this guide to find better ways to optimize inner loops.
+
+## Instruction selection and scheduling
+
+ARM provides several Software Optimization Guides (SWOG) for each CPU:
+- [Graviton2 - Neoverse-N1 SWOG](https://developer.arm.com/documentation/swog309707/a/)
+- [Graviton3 - Neoverse-V1 SWOG](https://developer.arm.com/documentation/PJDOC-466751330-9685/0101/)
+
+The SWOG documents for each instruction the latency (number of cycles it takes
+to finish the execution of an instruction) and the throughput (the number of
+similar instructions that can execute in parallel.) The SWOG also documents the
+execution units that can execute an instruction. The information provided by the
+SWOG is sometimes encoded in compilers (GCC and LLVM) under the form of specific
+tuning flags for different CPUs. Tuning flags allow compilers to produce a good
+instruction scheduling and a good instruction selection.
+
+LLVM has a tool [llvm-mca](https://www.llvm.org/docs/CommandGuide/llvm-mca.html)
+that allows software developers to see on their input assembly code how the
+compiler reasons about the execution of instructions by a specific processor.
+
+## Splitting data dependence chains
+
+To increase instruction level parallelism it is possible to duplicate
```

Comment:
Does this get done by default compiler optimizations? In general, it would be good to suggest how compiler can get these optimizations or if explicit user intervention is needed.